src/flow/src/expr/scalar.rs

```diff
@@ -151,6 +167,148 @@ pub enum ScalarExpr {
     },
 }
 
+/// a way to represent a scalar function that is implemented in Data Fusion
+#[derive(Debug, Clone, Serialize)]
+pub struct DfScalarFunction {
+    raw_fn: RawDfScalarFn,
+    // TODO(discord9): directly from datafusion expr
+    #[serde(skip)]
+    fn_impl: Arc<dyn PhysicalExpr>,
+    #[serde(skip)]
+    df_schema: Arc<datafusion_common::DFSchema>,
+}
+
+impl DfScalarFunction {
+    pub fn new(raw_fn: RawDfScalarFn, fn_impl: Arc<dyn PhysicalExpr>) -> Result<Self, Error> {
+        Ok(Self {
+            df_schema: Arc::new(raw_fn.input_schema.to_df_schema()?),
+            raw_fn,
+            fn_impl,
+        })
+    }
+
+    // TODO: add RecordBatch support
+    pub fn eval(&self, values: &[Value]) -> Result<Value, EvalError> {
+        // TODO(discord9): make cols all array length of one
+        let mut cols = vec![];
+        for (idx, typ) in self
```

Comment:
Please ensure that the number of columns in the schema is the same as the number of columns in the input row

Reply:
> Please ensure that the number of columns in the schema is the same as the number of columns in the input row

`DfRecordBatch::try_new` check that along many other things, I'm not sure whether I need double check that or not